File: Src/java/cql-to-elm/src/test/java/org/cqframework/cql/cql2elm/matchers/ConvertsToDecimalFrom.java
package org.cqframework.cql.cql2elm.matchers;

import org.hamcrest.Description;
import org.hamcrest.Factory;
import org.hamcrest.Matcher;
import org.hamcrest.TypeSafeDiagnosingMatcher;
import org.hl7.elm.r1.*;

import javax.xml.namespace.QName;

public class ConvertsToDecimalFrom extends TypeSafeDiagnosingMatcher<Expression> {
    private Object expectedArg;
    private ToDecimal expectedValue;

    public ConvertsToDecimalFrom(Integer i) {
        super();

        expectedArg = i;

        ObjectFactory of = new ObjectFactory();
        Literal integerLiteral = of.createLiteral()
                .withValueType(new QName("urn:hl7-org:elm-types:r1", "Integer"))
                .withValue(String.valueOf(i));

        QName expectedTypeName = new QName("urn:hl7-org:elm-types:r1", "Decimal");
        expectedValue = of.createToDecimal()
                .withOperand(integerLiteral);
    }

    public ConvertsToDecimalFrom(AliasRef a) {
        super();

        expectedArg = a;
        ObjectFactory of = new ObjectFactory();
        expectedValue = of.createToDecimal()
                .withOperand(a);
    }

    @Override
    protected boolean matchesSafely(Expression item, Description mismatchDescription) {
        if (! (item instanceof ToDecimal)) {
            mismatchDescription.appendText("had wrong ELM class type: ").appendText(item.getClass().getName());
            return false;
        }

        ToDecimal toDecimal = (ToDecimal) item;
        if (! expectedValue.equals(toDecimal)) {
            mismatchDescription.appendText("had wrong conversion: ").appendValue(toDecimal);
            return false;
        }

        return true;
    }

    @Override
    public void describeTo(Description description) {
        description.appendText("Conversion to Decimal w/ value: <")
                .appendValue(expectedArg);
    }

    @Factory
    public static <T> Matcher<Expression> convertsToDecimalFrom(Integer i) {
        return new ConvertsToDecimalFrom(i);
    }

    public static <T> Matcher<Expression> convertsToDecimalFromAlias(String alias) {
        AliasRef a = new ObjectFactory().createAliasRef().withName(alias);
        return new ConvertsToDecimalFrom(a);
    }
}

File: cql-to-elm/src/jvmTest/kotlin/org/cqframework/cql/cql2elm/matchers/ConvertsToDecimalFrom.kt
package org.cqframework.cql.cql2elm.matchers

import javax.xml.namespace.QName
import org.hamcrest.Description
import org.hamcrest.Factory
import org.hamcrest.Matcher
import org.hamcrest.TypeSafeDiagnosingMatcher
import org.hl7.elm.r1.AliasRef
import org.hl7.elm.r1.Expression
import org.hl7.elm.r1.ObjectFactory
import org.hl7.elm.r1.ToDecimal

class ConvertsToDecimalFrom : TypeSafeDiagnosingMatcher<Expression> {
    private val expectedArg: Any?
    private val expectedValue: ToDecimal

    constructor(i: Int?) : super() {
        expectedArg = i

        val of = ObjectFactory()
        val integerLiteral =
            of.createLiteral()
                .withValueType(QName("urn:hl7-org:elm-types:r1", "Integer"))
                .withValue(i.toString())

        expectedValue = of.createToDecimal().withOperand(integerLiteral)
    }

    constructor(a: AliasRef?) : super() {
        expectedArg = a
        val of = ObjectFactory()
        expectedValue = of.createToDecimal().withOperand(a)
    }

    @Suppress("ReturnCount")
    override fun matchesSafely(item: Expression, mismatchDescription: Description): Boolean {
        if (item !is ToDecimal) {
            mismatchDescription
                .appendText("had wrong ELM class type: ")
                .appendText(item.javaClass.simpleName)
            return false
        }

        if (expectedValue != item) {
            mismatchDescription.appendText("had wrong conversion: ").appendValue(item)
            return false
        }

        return true
    }

    override fun describeTo(description: Description) {
        description.appendText("Conversion to Decimal w/ value: <").appendValue(expectedArg)
    }

    @SuppressWarnings("MemberNameEqualsClassName")
    companion object {
        @JvmStatic
        @Factory
        fun convertsToDecimalFrom(i: Int): Matcher<Expression> {
            return ConvertsToDecimalFrom(i)
        }

        @JvmStatic
        fun convertsToDecimalFromAlias(alias: String): Matcher<Expression> {
            val a = ObjectFactory().createAliasRef().withName(alias)
            return ConvertsToDecimalFrom(a)
        }
    }
}
